File: 257-binary-tree-paths/257-binary-tree-paths.java

class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
    // List<String> answer = new ArrayList<String>();
    // if (root != null) searchBT(root, "", answer);
    // return answer;
        List<String> res=new ArrayList<>();
        Stack<TreeNode> nodes=new Stack<>();
        Stack<String> path=new Stack<>();
        nodes.push(root);
         if(root==null) return res;
        path.push("");
        while(!nodes.isEmpty())
        {
            TreeNode node=nodes.pop();
            String curStr=path.pop();
            if(node.left==null&&node.right==null) 
                res.add(curStr+node.val);
            if(node.left!=null) {
                nodes.push(node.left);
                path.push(curStr+node.val+"->");
            }
            if(node.right!=null)
            {
                nodes.push(node.right);
                path.push(curStr+node.val+"->");
            }           
        }           
        return res;
}
// private void searchBT(TreeNode root, String path, List<String> answer)
// {
//     if (root.left == null && root.right == null) answer.add(path + root.val);
   
//     if (root.left != null) searchBT(root.left, path + root.val + "->", answer);
    
//     if (root.right != null) searchBT(root.right, path + root.val + "->", answer);
//  }
}

File: 257-binary-tree-paths/257-binary-tree-paths.cpp
class Solution {
public:
//     void paths(TreeNode* root,string path,vector<string> &res)
//     {
//     if (root->left == NULL && root->right == NULL) res.push_back(path + to_string(root->val));
//     if (root->left != NULL) paths(root->left, path + to_string(root->val) + "->", res);
//     if (root->right != NULL) paths(root->right, path + to_string(root->val) + "->", res);
        
//     }
    
    vector<string> binaryTreePaths(TreeNode* root) {
      // vector<string> res;
      // if (root != NULL) paths(root,"",res);
      // return res;
        
         vector<string> list;
    stack<TreeNode*> sNode;
    stack<string> sStr;
    
    if(root==NULL) return list;
    sNode.push(root);
    sStr.push("");
    while(!sNode.empty()) {
        TreeNode* curNode=sNode.top(); sNode.pop();
        string curStr=sStr.top();sStr.pop();
        
        if(curNode->left==NULL && curNode->right==NULL) list.push_back(curStr+to_string(curNode->val));
        if(curNode->left!=NULL) {
            sNode.push(curNode->left);
            sStr.push(curStr+to_string(curNode->val)+"->");
        }
        if(curNode->right!=NULL) {
            sNode.push(curNode->right);
            sStr.push(curStr+to_string(curNode->val)+"->");
        }
    }
    return list;
    }
};